File: backend/internal/domain/models/product.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ProductCategory groups the standard products for easier filtering
type ProductCategory struct {
	ID        uuid.UUID  `json:"id"`
	Name      string     `json:"name"`
	ParentID  *uuid.UUID `json:"parent_id,omitempty"`
	CreatedAt time.Time  `json:"created_at"`
}

// Product represents a sellable catalog item
type Product struct {
	ID          uuid.UUID  `json:"id"`
	CategoryID  *uuid.UUID `json:"category_id,omitempty"`
	SKU         *string    `json:"sku,omitempty"` // Mapped from 'code' in DB
	Name        string     `json:"name"`
	Description *string    `json:"description,omitempty"`
	Unit        *string    `json:"unit,omitempty"`
	Price       float64    `json:"price"` // Mapped from 'base_price' in DB
	IsActive    bool       `json:"is_active"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// DealItem represents the junction appending products into a deal 
type DealItem struct {
	ID          uuid.UUID `json:"id"`
	DealID      *uuid.UUID `json:"deal_id,omitempty"`
	ProductID   uuid.UUID `json:"product_id"`
	Name        string    `json:"name"`
	Quantity    float64   `json:"quantity"` // Changed to float for fractions if needed
	Unit        string    `json:"unit"`     // pcs, dus, crate
	UnitPrice   float64   `json:"unit_price"`
	Discount    float64   `json:"discount"`
	Subtotal    float64   `json:"subtotal"` // Formula: (UnitPrice * Quantity) - Discount
	Notes       *string   `json:"notes,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
}

File: mobile/lib/features/products/domain/entities/product.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'product.freezed.dart';
part 'product.g.dart';

@freezed
abstract class Product with _$Product {
  const factory Product({
    required String id,
    @JsonKey(name: 'category_id') String? categoryId,
    required String name,
    String? description,
    required double price,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    String? sku,
    @JsonKey(name: 'created_at') String? createdAt,
    @JsonKey(name: 'updated_at') String? updatedAt,
  }) = _Product;

  factory Product.fromJson(Map<String, dynamic> json) =>
      _$ProductFromJson(json);
}
